### PR TITLE
Notify `systemd(1)` about start-up completion and other service status changes

### PR DESCRIPTION
**Pull request by basil**
Refactors lifecycle-related notifications into a set of common methods in `Lifecycle`, then adds a new `SystemdLifecycle` that implements support for [`sd_notify(3)`](https://www.freedesktop.org/software/systemd/man/sd_notify.html). This allows `systemd` to have awareness of when Jenkins has finished starting up and allows `systemd` to correctly mark the service as failed if the Java process starts but Jenkins startup does not complete (e.g., due to a JCasC error).

I tested this with the following settings:

```
[Service]
Type=notify
NotifyAccess=main
```

Manual testing done:

- Started Jenkins with `systemctl start jenkins`; verified the command blocked until the service had fully started.
- Stopped Jenkins with `/safeExit`; verified the service was marked as `inactive (dead)` with `Status: "Jenkins stopped"`.
- Restarted Jenkins with `/safeRestart`; verified the service had `Status: "Restart in 10 seconds"`, then `Active: activating (start)`, then finally `Active: active (running)` after restarting.
- Updated a plugin and checked "Restart Jenkins when installation is complete and no jobs are running"; verified the service had `Status: "Restart in 10 seconds"`, then `Active: activating (start)`, then finally `Active: active (running)` after restarting.
- Ran the Molecule tests from jenkinsci/packaging#266 successfully against this PR.

### Proposed changelog entries

Startup completion notification is available with `systemd(1)`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (basil) [inline core/src/main/java/jenkins/model/Jenkins.java:4544]: This gives us fancy status notifications in `systemctl status`:

```
$ sudo systemctl status jenkins
● jenkins.service - Jenkins Continuous Integration Server
     Loaded: loaded (/lib/systemd/system/jenkins.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/jenkins.service.d
             └─override.conf
     Active: active (running) since Sat 2022-01-29 16:52:16 PST; 1min 4s ago
   Main PID: 169136 (java)
     Status: "Restart in 10 seconds"
      Tasks: 59 (limit: 38237)
     Memory: 2.9G
     CGroup: /system.slice/jenkins.service
             └─169136 /usr/bin/java -Djava.awt.headless=true -jar /usr/share/java/jenkins.war --webroot=/var/cache/jenkins/war --httpPort=8080
```

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by basil** [comment]
@maxime-michel Update `/lib/init/init-d-script` to replace the first line `#!/bin/sh` with `#!/bin/sh -x` and run `systemctl daemon-reload` and `systemctl restart jenkins`.

Then file a Jira issue with:

- Your OS information (e.g. Ubuntu 20.04.3 LTS x86_64)
- The output of `dpkg -s sysvinit-utils | grep Version`
- The contents of your `/etc/default/jenkins`
- The debug output from `journalctl -u jenkins -b`. It should look like this:

```
+ __init_d_script_name=/etc/init.d/jenkins
+ shift
+ . /lib/lsb/init-functions
[…]
```

**Comment by basil** [comment]
@maxime-michel See also [JENKINS-67724](https://issues.jenkins.io/browse/JENKINS-67724) which you may well be hitting (fixed but unreleased; workaround described in Jira).

**Pull request merged**
